client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -137,50 +155,116 @@ public void start(Listener listener) {
         @Override
-        @Synchronized
         public void shutdown() {
             Preconditions.checkState(started.compareAndSet(true, false));
             if (this.scheduledExecutor != null) {
                 this.scheduledExecutor.shutdownNow();
             }
         }
 
-        // The controller discovery API invoker.
+        @Override
+        @Synchronized
+        public void refresh() {
+            // Refresh is called as hints when gRPC detects network failures.
+            // We don't want to repeatedly attempt discovery; following logic will limit discovery on failures to
+            // once every FAILURE_RETRY_TIMEOUT_SECONDS seconds. Also we want to trigger discovery sooner on failures.
+            if (started.get()) {
+                if (this.scheduledFuture != null &&
+                        !this.scheduledFuture.isDone() &&
+                        this.scheduledFuture.getDelay(TimeUnit.SECONDS) > FAILURE_RETRY_TIMEOUT_SECONDS) {
+                    // Cancel the existing schedule and advance the discovery process.
+                    this.scheduledFuture.cancel(true);
+                    this.scheduledFuture = this.scheduledExecutor.schedule(
+                            this::getControllers, FAILURE_RETRY_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+                }
+            }
+        }
+
+        /**
+         * The controller discovery API invoker.
+         * This refreshes the list of controller addresses to be used by the gRPC transport.
+         * The discovery process will be rescheduled in the end after a delay which is calculated based on whether
+         * the controller addresses have been fetched successfully or not.
+         */
+        @Synchronized
         private void getControllers() {
+            log.info("Attempting to refresh the controller server endpoints");
             try {
-                final ServerResponse controllerServerList =
-                        this.client.getControllerServerList(ServerRequest.getDefaultInstance());
-                final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder()
-                        .addAll(controllerServerList.getNodeURIList()
-                                .stream()
-                                .map(node ->
-                                        new ResolvedServerInfo(
-                                                new InetSocketAddress(node.getEndpoint(), node.getPort())))
-                                .collect(Collectors.toList()))
-                        .build();
+                final ResolvedServerInfoGroup serverInfoGroup;
+                if (this.enableDiscovery) {
+                    // Make an RPC call to the bootstrapped controller servers to fetch all active controllers.
+                    final ServerResponse controllerServerList =
+                            this.client.getControllerServerList(ServerRequest.getDefaultInstance());
```

Comment:
We could make this non-blocking if we really want to optimize the thread usage for this, especially if using an external executor. Since we make only one call at a time and have a dedicated thread for this, we don't need to do this. I'll file an issue for this since this is non-trivial, especially for DNS lookups. The current java apis for DNS lookups are blocking, and we may have to explore using 3rdparty libraries for a non-blocking DNS client.